packages/realm-app-importer/src/types.ts

```diff
@@ -0,0 +1,231 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+type ErrorResponse = {
+  error: string;
+};
+
+export function isErrorResponse(response: unknown): response is ErrorResponse {
+  if (typeof response === "object" && response !== null) {
+    assertObject(response);
+    return typeof response.error === "string";
+  } else {
+    return false;
+  }
+}
+
+export type LoginResponse = {
+  access_token: string;
+  refresh_token: string;
+};
+
+export function assertLoginResponse(response: unknown): asserts response is LoginResponse {
+  assertObject(response);
+  if (typeof response.access_token !== "string" && typeof response.refresh_token !== "string") {
+    throw new Error("Expected a LoginResponse");
+  }
+}
+
+export type App = {
+  _id: string;
+  client_app_id: string;
+  name: string;
+  location: string;
+  deployment_model: string;
+  domain_id: string;
+  group_id: string;
+  last_used: number;
+  last_modified: number;
+  product: string;
+  environment: string;
+};
+
+export type AppResponse = App;
+
+export type AppsResponse = Array<App>;
+
+export type ProfileResponse = {
+  roles: Array<{ role_name: string; group_id: string }>;
+};
+
+export function assertObject(response: unknown): asserts response is Record<string, unknown> {
+  if (typeof response !== "object" || response === null) {
+    throw new Error("Expected an object");
+  }
+}
+
+export function assertArray(response: unknown): asserts response is Array<unknown> {
+  if (!Array.isArray(response)) {
+    throw new Error("Expected an array");
+  }
+}
+
+export function assertAppResponse(response: unknown): asserts response is AppResponse {
+  assertObject(response);
+  if (typeof response.client_app_id !== "string" || typeof response._id !== "string") {
+    throw new Error("Expected an AppResponse");
+  }
+}
+
+export function assertAppsResponse(response: unknown): asserts response is AppsResponse {
+  assertArray(response);
+  response.every(assertAppResponse);
+}
+
+export type DeploymentDraft = {
+  _id: string;
+  user_id: string;
+  app: App;
+};
+
+export type DeploymentDraftResponse = DeploymentDraft;
+
+export function assertDeploymentDraft(value: unknown): asserts value is DeploymentDraftResponse {
+  assertObject(value);
+  if (typeof value._id !== "string" || typeof value.user_id !== "string" || typeof value.app !== "object") {
+    throw new Error("Expected a DeploymentDraftResponse");
+  }
```

Comment:
```suggestion
  if (typeof value._id !== "string" || typeof value.user_id !== "string") {
    assertObject(value.app);
    throw new Error("Expected a DeploymentDraftResponse");
  }
```

Reply:
In practice yes. I didn't want to make it too dependent on the baas implementation 🙈
Ideally, I would have preferred generating all of this from an open api spec 🤔